### PR TITLE
Part II, Lesson 4

### LINKED ISSUE #15: Part II, Lesson 4
This lesson will introduce object-to-object interactions.

1. Introduce a new object, a ball.
2. Give it some momentum, and make it bounce against the edges of the playfield.
3. Then, make it so the ball bounces against the paddle as well.
4. If judged appropriate, it can be shown how the ball can clip against the paddle's vertical edges, and this can then be accounted for.

### PR DESCRIPTION
**Pull request by eievui5**
Closes #15

### CODE REVIEW

**Comment by eievui5** [comment]
At the moment this only implements ball movement, not bouncing.

**Comment by ISSOtm** [comment]
The current code looks fine to me.

**Comment by avivace** [comment]
What are we missing here? The text part? @eievui5 do you have any update on this?

**Comment by eievui5** [comment]
We're missing the ball bouncing against the walls and the paddle, and the text.

**Comment by eievui5** [comment]
I'd like some feedback on the bouncing code. The pixel -> tilemap function might be too complex for the tutorial but I'm not sure how else to write it. Otherwise, I think this is pretty understandable :)

File: unbricked/collision/main.asm [hunk 1]
```diff
@@ -114,6 +114,68 @@ WaitVBlank2:
 	ld [_OAMRAM + 4], a
 ; ANCHOR_END: momentum
 
+; ANCHOR: tile-collision
+BounceOnTop:
+	; Remember to offset the OAM position!
+	; (8, 16) in OAM coordinates is (0, 0) on the tilemap.
```
Comment (ISSOtm) [inline unbricked/collision/main.asm:120]:
```suggestion
	; (8, 16) in OAM coordinates is (0, 0) on the screen.
```

File: unbricked/collision/main.asm [hunk 2]
```diff
@@ -61,9 +61,9 @@ ClearOam:
 	ld [hli], a
 ; ANCHOR: init
 	; Now initialize the ball sprite
-	ld a, 64 + 16
+	ld a, 100 + 16
```
Comment (ISSOtm) [inline unbricked/collision/main.asm:64]: How about using the `hardware.inc` constants instead?

File: unbricked/collision/main.asm [hunk 3]
```diff
@@ -147,6 +209,57 @@ Right:
...
+	ld bc, $9800
+	add hl, bc
+	ret
+; ANCHOR_END: get-tile
+
+; ANCHOR: is-wall-tile
+; @param a: tile ID
+; @return z: set if a is a wall.
+IsWallTile:
+	cp a, $00
```
Comment (ISSOtm) [inline unbricked/collision/main.asm:245]: What are those IDs about?

**Review by ISSOtm — CHANGES_REQUESTED** [review]
Alright, the logic seems correct, I just hope we won't start overflowing VBlank time... ><

File: unbricked/collision/main.asm (hunk 3 above)
Comment (eievui5) [reply unbricked/collision/main.asm:245]: Those are all of the tiles that border the playfield + the bricks.

File: unbricked/collision/main.asm (hunk 1 above)
Comment (eievui5) [reply unbricked/collision/main.asm:120]: This isn’t accurate because the function is converting a tilemap position not a… screen position?

File: unbricked/collision/main.asm (hunk 3 above)
Comment (ISSOtm) [reply unbricked/collision/main.asm:245]: Ah, right. How about introducing carry comparisons, then?

File: unbricked/collision/main.asm (hunk 1 above)
Comment (ISSOtm) [reply unbricked/collision/main.asm:120]: We are converting to parameters for `GetTileByPixel`, which appears take screen positions?

Comment (eievui5) [reply unbricked/collision/main.asm:120]: Oh yeah. Okay, that makes sense

File: unbricked/collision/main.asm (hunk 3 above)
Comment (eievui5) [reply unbricked/collision/main.asm:245]: Well because you’ll notice they’re not continuous and i don’t think it’s a good idea to reorder them at this point. But maybe it’s worth it for carry

Comment (ISSOtm) [reply unbricked/collision/main.asm:245]: I actually did a double-take to try and find a discontinuity, because I didn't expect *you* of all people to pull off a `GetForestTreeFrame` :P

So I think it would help, here.

Comment (eievui5) [reply unbricked/collision/main.asm:245]: Okay. My only issue is that this would be a “breaking change” for anyone who followed the tutorial in the past and is catching up on the new chapter

File: unbricked/collision/main.asm [hunk 4]
```diff
@@ -0,0 +1,581 @@
...
+; ANCHOR_END: init
+
+	; Turn the LCD on
+	ld a, LCDCF_ON | LCDCF_BGON | LCDCF_OBJON
+	ld [rLCDC], a
+
+	; During the first (blank) frame, initialize display registers
+	ld a, %11100100
+	ld [rBGP], a
+	ld a, %11100100
```
Comment (Rangi42) [inline unbricked/collision/main.asm:88]: @ISSOtm If you're okay with the redundant `ld`, I think the `IsWallTile` `cp`s are fine in comparison. (They're trivial to add/remove individual tile IDs from.)

File: unbricked/collision/main.asm (hunk 3 above)
Comment (ISSOtm) [reply unbricked/collision/main.asm:245]: Oh, I was only saying to use a carry comparison for the contiguous IDs, and tack on the final `cp $09`.

File: unbricked/collision/main.asm (hunk 4 above)
Comment (ISSOtm) [reply unbricked/collision/main.asm:88]: Didn't we discuss that this one should be removed?

Comment (eievui5) [reply unbricked/collision/main.asm:88]: iirc ISSO wanted to remove that `ld` actually. While being able to remove the IDs from IsWallTile, I can just reorder the tiles such that the last two are the bricks, and then in the next chapter reduce the constant by two. *Or*, just remove the bricks from the list until then.

Comment (ISSOtm) [reply unbricked/collision/main.asm:88]: I think removing the bricks is acceptable for now. Especially since we won't be able to just rely on the tile ID for the bounce direction anymore.

File: unbricked/collision/main.asm (hunk 3 above)
Comment (eievui5) [reply unbricked/collision/main.asm:245]: What about the missing `$03`? did you miss that?

Comment (ISSOtm) [reply unbricked/collision/main.asm:245]: Yep!

Comment (eievui5) [reply unbricked/collision/main.asm:245]: Without reordering these, I don't see a point in using carry, and I'd rather leave it like this.

Comment (eievui5) [reply unbricked/collision/main.asm:245]: Yes, it would be a bit faster, but interleaving `a != 3 && a <= 7 || a == 9` is kinda needless.

File: unbricked/collision/main.asm (hunk 2 above)
Comment (eievui5) [reply unbricked/collision/main.asm:64]: Using hardware.inc for what? You mean like `SCRN_Y / 2`?

**Comment by eievui5** [comment]
The text and code is now complete.

**Review by avivace — COMMENTED** [review]
Wow, this actually reads very smoothly! I just left some minor comments and suggestions and will go over it again later.

File: src/part2/collision.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
+# Collision
+
+Being able to move around is great, but there's still one object we need for this game: a ball!
+Just like with the paddle, the first step is to create a graphic for the ball and load it into VRAM.
```
[the remark this answers is not in the document]
Comment (eievui5) [reply src/part2/collision.md:4]: "create a tile" could work

File: unbricked/collision/main.asm [hunk 6]
```diff
@@ -0,0 +1,610 @@
...
+	; Now let's compare the X positions of the objects to see if they're touching.
+	ld a, [_OAMRAM + 1]
+	ld b, a
+	ld a, [_OAMRAM + 5]
+	add a, 16
+	cp a, b
+	jp c, PaddleBounceDone
+	sub a, 16 + 8
+	cp a, b
+	jp nc, PaddleBounceDone
```
Comment (ISSOtm) [inline unbricked/collision/main.asm:199]: I think the X position check is easier to understand and explain if done the other way around:
```suggestion
	ld a, [_OAMRAM + 5] ; Ball's X position.
	ld b, a
	ld a, [_OAMRAM + 1] ; Paddle's X position.
	sub a, 8
	cp a, b
	jp c, PaddleBounceDone
	add a, 8 + 16 ; 8 to undo, 16 as the width.
	cp a, b
	jp nc, PaddleBounceDone
```

I will also make a diagram to accompany the textual explanation, once a solution is agreed upon.

File: src/part2/collision.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+The Y position's check is simple, since our paddle is flat.
+However, the X position has two checks which widen the area the ball can bounce on.
+First we add 16 to the ball's position; if the ball is more than 16 pixels to the right of the paddle, it shouldn't bounce.
+Then we undo this by subtracting 16, and while we're at it, subtract another 8 pixels; if the ball is more than 8 pixels to the left of the paddle, it shouldn't bounce.
+
+You might be wondering why we checked 16 pixels to the right but only 8 pixels to the left.
+Remember that OAM positions represent the upper-*left* corner of a sprite, so the center of our paddle is actually 4 pixels to the right of the position in OAM.
+When you consider this, we're actually checking 12 pixels out on either side from the center of the paddle.
+12 pixels might seem like a lot, but it gives some tolerance to the player in case their positioning is off.
+If you'd prefer to make this easier or more difficult, feel free to adjust the values!
```
Comment (ISSOtm) [inline src/part2/collision.md:107]:
```suggestion
::: note

You might be wondering why we checked 16 pixels to the right but only 8 pixels to the left.
Remember that OAM positions represent the upper-*left* corner of a sprite, so the center of our paddle is actually 4 pixels to the right of the position in OAM.
When you consider this, we're actually checking 12 pixels out on either side from the center of the paddle.

12 pixels might seem like a lot, but it gives some tolerance to the player in case their positioning is off.
If you'd prefer to make this easier or more difficult, feel free to adjust the values!

:::
```

**Review by ISSOtm — CHANGES_REQUESTED** [review]
Great work! I only have some clean-up suggestions.

Additionally, I would prefer if the Z flag was capitalised, for consistency with the "flags" lesson. (The rationale there was to avoid confusion with the c register.)

And maybe we could link back to some previous lessons, if they are relevant.

File: unbricked/collision/main.asm (hunk 2 above)
Comment (ISSOtm) [reply unbricked/collision/main.asm:66]: I was thinking of [`OAM_Y_OFS` and `OAM_X_OFS`](https://github.com/gbdev/hardware.inc/blob/225e671a9eb1f81c90e0cc17d29e59c6983657db/hardware.inc#L1040-L1041).

Maybe we can introduce those in an intermediate "cleanup" lesson, where we would introduce constants, and split `main.asm` into several files? (It's getting quite bulky.)

**Comment by zlago** [comment]
> Additionally, I would prefer if the Z flag was capitalised, for consistency with the "flags" lesson. (The rationale there was to avoid confusion with the c register.)

may i ask why you wont state that theyre called `Z`/`C`, but that the tutorial will refer to them as `zf`/`cf` (or `z`/`cy`) to avoid confusion with registers?

**Comment by ISSOtm** [comment]
I just went with common practice. Using one of these names would probably help (and `cf`/`cy` might help avoid the `rrca`/`rra` confusion), but I don't want to have to go back to previous lessons as well :smiling_face_with_tear:

File: src/part2/collision.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+{{#include ../../unbricked/collision/main.asm:oam}}
+```
+
+As the ball bounces around the screen its momentum will change, sending it in different directions.
+Let's create two new variables to track the ball's momentum in the two axis: `wBallMomentumX` and `wBallMomentumY`.
+```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/collision/main.asm:ram}}
+{{#include ../../unbricked/collision/main.asm:ram}}
+```
+
+We'll need to initialize these, so let's so do right after we write the ball to OAM.
```
[the remark this answers is not in the document]
Comment (avivace) [reply src/part2/collision.md:27]: I think this lesson is "loaded" enough, let's do that in one of the following lessons.

File: src/part2/collision.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
+# Collision
+
+Being able to move around is great, but there's still one object we need for this game: a ball!
+Just like with the paddle, the first step is to create a tile for the ball and load it into VRAM.
+
```
[the remark this answers is not in the document]
Comment (avivace) [reply src/part2/collision.md:5]: Agree

File: unbricked/collision/main.asm (hunk 6 above)
Comment (eievui5) [reply unbricked/collision/main.asm:199]: A diagram would be great, I considered doing it in text but figured an explanation would be enough

File: unbricked/collision/main.asm (hunk 2 above)
Comment (eievui5) [reply unbricked/collision/main.asm:66]: that works for me

File: src/part2/collision.md [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+{{#include ../../unbricked/collision/main.asm:copy-ball}}
+```
+
+In addition, we need to initialize an entry in OAM, right after where we initialize the paddle.
+```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/collision/main.asm:oam}}
+{{#include ../../unbricked/collision/main.asm:oam}}
+```
+
+As the ball bounces around the screen its momentum will change, sending it in different directions.
+Let's create two new variables to track the ball's momentum in the two axis: `wBallMomentumX` and `wBallMomentumY`.
```
[the remark this answers is not in the document]
Comment (eievui5) [reply src/part2/collision.md:24]: Honestly I prefer the original wording ("in each axis"); "axes" is quite uncommon.

File: src/part2/collision.md [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+You might want to compile your game again to see what this does.
+If you do, you should see the ball moving around, but it will just go through the walls and offscreen.
+
+We need to add collision with the walls so that the ball can bounce around.
+There's some complexity to doing this, so we're going to make use of two functions.
+
+Please do not get stuck on the details of this next function, as it uses some techniques and instructions we haven't discussed yet.
+The basic idea is that it converts the position of the sprite to a location on the tilemap.
+This way, we can check which tile our ball is touching so that we know when to bounce!
```
Comment (ISSOtm) [inline src/part2/collision.md:54]:
```suggestion
::: tip

Please do not get stuck on the details of this next function, as it uses some techniques and instructions we haven't discussed yet.
The basic idea is that it converts the position of the sprite to a location on the tilemap.
This way, we can check which tile our ball is touching so that we know when to bounce!

:::

```

File: src/part2/collision.md [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+The next function is called `IsWallTile`, and it's going to contain a list of tiles which the ball can bounce off of.
+This should be easier to understand!
+```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/collision/main.asm:is-wall-tile}}
+{{#include ../../unbricked/collision/main.asm:is-wall-tile}}
+```
+
+This function might look a bit strange at first.
+Instead of returning its result in a *register*, like `a`, it returns it in [a *flag*](../part1/operations.md#flags): `Z`!
+If at any point a tile matches, the function has found a wall and exits with `Z` set.
+If the target tile ID (in `a`) matches one of the wall tile IDs, the corresponding `cp` will leave `Z` set; if so, we return immediately (via `ret Z`), with `Z` set.
```
Comment (ISSOtm) [inline src/part2/collision.md:68]:
```suggestion
If the target tile ID (in `a`) matches one of the wall tile IDs, the corresponding `cp` will leave `Z` set; if so, we return immediately (via `ret z`), with `Z` set.
```

File: src/part2/collision.md [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+{{#include ../../unbricked/collision/main.asm:tile-collision}}
+```
+
+That was a lot, but now the ball bounces around your screen!
+There's just one last thing to do before this chapter is over, and thats ball-to-paddle collision.
+
+## Paddle bounce
+
+Unlike with the tilemap, there's no position conversions to do here, just straight comparisons.
+However, we'll need to cover a new concept: the *carry* flag.
```
Comment (ISSOtm) [inline src/part2/collision.md:97]:
```suggestion
However, for these, we will need [the *carry* flag](../part1/operations.md#flags).
```

File: src/part2/collision.md [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+Carry is represented by a `C`, like how zero is a `Z`, but don't get it confused with the `c` register!
```
Comment (ISSOtm) [inline src/part2/collision.md:98]:
```suggestion
The carry flag is notated as `C`, like how the zero flag is notated as `Z`, but don't confuse it with the `c` register!
```

File: src/part2/collision.md [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+
+Just like `Z`, you can use it to conditionally jump.
+However, while `Z` is used to check if two numbers are equal, `C` can be used to check if numbers are greater than or less than each other.
+For example, `cp a, b` sets `C` if `a < b`, and does not set `C` if `a >= b`.
+(If you want to check `a <= b` or `a > b`, you can use `Z` and `C` in tandem with two `jp` instructions)
```
Comment (ISSOtm) [inline src/part2/collision.md:103]:
```suggestion
::: tip A refresher on comparisons

Just like `Z`, you can use the carry flag to jump conditionally.
However, while `Z` is used to check if two numbers are equal, `C` can be used to check if a number is greater than or smaller than another one.
For example, `cp a, b` sets `C` if `a < b`, and clears it if `a >= b`.
(If you want to check `a <= b` or `a > b`, you can use `Z` and `C` in tandem with two `jp` instructions.)

:::
```

**Review by ISSOtm — CHANGES_REQUESTED** [review]
Some minor changes, and then I'll have to make the explanation diagram for the last paragraph.

File: src/part2/collision.md (hunk 12 above)
Comment (eievui5) [reply src/part2/collision.md:68]: I don't like this inconsistency

**Comment by eievui5** [comment]
Go ahead :)

File: src/part2/collision.md (hunk 12 above)
Comment (ISSOtm) [reply src/part2/collision.md]: It would be inconsistent with the code.

**Review by avivace — APPROVED** [review]

**Comment by ISSOtm** [comment]
Here are screenshots of the diagram:

image

image

Maybe it should be clarified that we are considering the position of the ball's top-left pixel, though it is highlighted in the above. And, speaking of, going 16 pixels to the right doesn't seem to make much sense given how wide the paddle appears to be. I dunno.

**Comment by eievui5** [comment]
> going 16 pixels to the right doesn't seem to make much sense given how wide the paddle appears to be
I’m not sure how much you’ve played around with the physics but i think this has the best feel to it. Any smaller and the player would probably be frustrated by the small paddle size. It might be worth doubling the width of the paddle with another sprite but i don’t think that’s necessary.

**Comment by ISSOtm** [comment]
Then maybe we should offset the hitbox a bit more to the left, so it's symmetrical?

**Comment by eievui5** [comment]
I'd rather just go edit the ball graphic to be one pixel over to the right. Could you update your diagram to reflect that?

**Comment by ISSOtm** [comment]
Sure, it's just a PNG, you can edit it in `assets/part2`.

**Comment by eievui5** [comment]
Thought it was an svg :) should be good now

**Comment by ISSOtm** [comment]
Alright, that looks good to me. Should we add a "bonus" heading, suggesting how to make the ball bounce when its bottom pixel touches the paddle's? We should suggest "you only need a single instruction!" (`sub a, 6`).

**Comment by eievui5** [comment]
Maybe. I like that it kinda sinks in. If we wanna add that in at all then maybe it should just be part of the tutorial itself.

**Comment by ISSOtm** [comment]
I think it would be beneficial to encourage experimentation, especially if it's just a single instruction.

File: src/part2/collision.md [hunk 16]
```diff
@@ -220,3 +220,21 @@ When you consider this, we're actually checking 12 pixels out on either side fro
...
+<details><summary>Answer:</summary>
+
+```diff linenos,start={{#line_no_of "" ../../unbricked/collision/main.asm:paddle-bounce}}
+	ld a, [_OAMRAM]
+	ld b, a
+	ld a, [_OAMRAM + 4]
++	sub a, 6
+	cp a, b
+```
+
```
Comment (ISSOtm) [inline src/part2/collision.md:239]:
```suggestion

Alternatively, you can add `add a, 6` just after `ld a, [_OAMRAM]` instead.

In both cases, try playing with that `6` value, see what feels right!

```

**Review by ISSOtm — CHANGES_REQUESTED** [review]
Great!

**Review by ISSOtm — APPROVED** [review]

File: src/part2/collision.md [hunk 17]
```diff
@@ -0,0 +1,244 @@
...
+That was a lot, but now the ball bounces around your screen!
+There's just one last thing to do before this chapter is over, and thats ball-to-paddle collision.
+
+## Paddle bounce
+
+Unlike with the tilemap, there's no position conversions to do here, just straight comparisons.
+However, for these, we will need [the *carry* flag](../part1/operations.md#flags).
+The carry flag is notated as `C`, like how the zero flag is notated as `Z`, but don't confuse it with the `c` register!
+
+::: tip A refresher on comparisons
```
Comment (avivace) [inline src/part2/collision.md:105]: minor: I really like those boxes!

File: src/part2/collision.md [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+
+```diff linenos,start={{#line_no_of "" ../../unbricked/collision/main.asm:paddle-bounce}}
+	ld a, [_OAMRAM]
+	ld b, a
+	ld a, [_OAMRAM + 4]
++	sub a, 6
+	cp a, b
+```
+
+Alternatively, you can add `add a, 6` just after `ld a, [_OAMRAM]` instead.
```
Comment (avivace) [inline src/part2/collision.md:240]: Doesn't Alternatively already says what "instead" does? We should pick one only.

**Review by avivace — CHANGES_REQUESTED** [review]

File: src/part2/collision.md [hunk 19]
```diff
@@ -0,0 +1,244 @@
...
+
+You might want to compile your game again to see what this does.
+If you do, you should see the ball moving around, but it will just go through the walls and offscreen.
+
+We need to add collision with the walls so that the ball can bounce around.
+There's some complexity to doing this, so we're going to make use of two functions.
+
+::: tip
+
+Please do not get stuck on the details of this next function, as it uses some techniques and instructions we haven't discussed yet.
```
[the remark this answers is not in the document]
Comment (eievui5) [reply src/part2/collision.md:54]: There is *so much* to explain here that it would take an entire chapter to do so. I don't see any way to explain this without wasting a lot of the reader's time.

File: src/part2/collision.md [hunk 20]
```diff
@@ -0,0 +1,244 @@
...
+This can get pretty confusing, but luckily we only have two objects to keep track of.
+In the future, we'll go over a much easier way to use OAM.
+```rgbasm,linenos,start={{#line_no_of "" ../../unbricked/collision/main.asm:momentum}}
+{{#include ../../unbricked/collision/main.asm:momentum}}
+```
+
+You might want to compile your game again to see what this does.
+If you do, you should see the ball moving around, but it will just go through the walls and offscreen.
+
+We need to add collision with the walls so that the ball can bounce around.
```
[the remark this answers is not in the document]
Comment (eievui5) [reply src/part2/collision.md:49]: I feel like "add collision with the walls" makes sense (the ball will collide with the walls) whereas "add collision detection with the walls" is more confusing (the walls will be used to detect collision?)

File: src/part2/collision.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (ISSOtm) [reply src/part2/collision.md:54]: All things considered, I think it would make sense to discuss this function in a chapter of its own, since there is so much to discuss.

Considering the way Part II is written, I think we should explain it after introducing it, so that all of the dry concepts and explanations will at least be in a context and goal already known to the reader.

Thus, I think we should dedicate the lesson following this one to that function. Does that seem reasonable?

File: src/part2/collision.md (hunk 18 above)
Comment (ISSOtm) [reply src/part2/collision.md]: I don't think it does, as an alternative could be to only add code. So I think the redundancy here helps clarify the intent.

File: src/part2/collision.md [hunk 21]
```diff
@@ -0,0 +1,244 @@
...
+There's some complexity to doing this, so we're going to make use of two functions.
```
[the remark this answers is not in the document]
Comment (ISSOtm) [reply src/part2/collision.md]: How about "since the code is somewhat lenghty, we will split some of it out into two functions"?

[the remark this answers is not in the document]
Comment (eievui5) [reply src/part2/collision.md]: I’ve already changed this to mention that the code needs to be *repeated* a few times.

[the remark this answers is not in the document]
Comment (ISSOtm) [reply src/part2/collision.md]: I blame GitHub Mobile's UI not showing the changes you made since. :( Sorry!

File: src/part2/collision.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (avivace) [reply src/part2/collision.md:49]: > I agree with Evie there. Plus, collision is not just detection, it is also reaction.
I'm not following you. For you _adding collision_ is okay , even if it literally doesn't mean anything in plain english but "adding collision detection" is wrong because we should include reaction too? Try to google "collision detection gamedev". In almost every single result to add the "detection" of something also implies (or it's the natural prerequisite) for reacting to it.

> "add collision detection and response"
I don't see how this is complex, especially because this is what we are doing and presenting anyway, so how can the fact that "it's too complex" be a problem at all?

File: src/part2/collision.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (avivace) [reply src/part2/collision.md:54]: If there is so much to discuss, and if it requires _an entire_ chapter, why exactly are we pushing this to the reader without discussing it beforehand? I don't see any advantage in doing so, except adding complexity in how are the concept presented. What I can propose is to actually write "this chapter that will explain this block" and then we merge/add it here. Eventually maybe this chapter could be split in 2 if it's too long, but I believe that the we definitely exaggerated in the "goal" for this chapter by pushing too much "unexplained" or "ignore this , will be explained" later stuff

File: src/part2/collision.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (eievui5) [reply src/part2/collision.md:49]: It's not really about the use of "detection" so much as the "with the walls" part. If we want to make it say "collision detection", it would have to look like this:
> We need to add collision detection so that the ball can bounce around.
which works.

File: src/part2/collision.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (eievui5) [reply src/part2/collision.md:54]: It's simply because all the math and instructions behind this would completely throw off the pacing of the tutorial. Our primary concern at the moment is teaching the logic of assembly and making a game.

I don't think anyone wants to start off "Hello, world!" by explaining all the intricacies of file descriptors and pipes, for example. I would consider this function to be "`#include <stdio.h>`". It would be nice to explain more, but now is not the time.

[the remark this answers is not in the document]
Comment (ISSOtm) [reply src/part2/collision.md:54]: Showing results and explaining them later has been the entire guiding principle behind Part I, since the gb-asm-tutorial-old showed that the "explain everything first, show results at the end" strategy was ineffective.

I think it also helps making the tutorial feeling more cohesive; think of it as a plot thread/cliffhanger?

Maybe we should reword this info box, however. How about:

> This function uses a lot of concepts we haven't talked about yet.
> We will examine and explain it in the next section; for now, let's assume it does what we want it to.

[the remark this answers is not in the document]
Comment (eievui5) [reply src/part2/collision.md:54]: "next section" being part 3?

[the remark this answers is not in the document]
Comment (ISSOtm) [reply src/part2/collision.md:54]: I meant "next lesson", don't mind me.

**Review by avivace — APPROVED** [review]
I don't agree with the suggested order of presenting things (let alone pushing entire functions/sections of code to the reader) but it's merely an order issue, it can be re-examined later when such explanations are added.

Everything else looks very good
